src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -100,7 +99,7 @@ internal WebApplicationBuilder(WebApplicationOptions options, Action<IHostBuilde
             Host = new ConfigureHostBuilder(hostContext, Configuration, Services);
             WebHost = new ConfigureWebHostBuilder(webHostContext, Configuration, Services);
 
-            Services.AddSingleton<IConfiguration>(Configuration);
+            Services.AddSingleton<IConfiguration>(_ => Configuration);
```

Comment:
Is there a functional difference with this change?

Reply:
By changing the service descriptor from an instance to a factory, this gives the host ownership and now it disposes the ConfigurationManager. What's interesting is HostBuilder itself was using the same trick for its ConfigurationRoot and that no longer gets disposed, but it doesn't matter because ConfigurationManager took ownership of all the providers except the ChainedConfigurationSource to itself.

Reply:
> but it doesn't matter because ConfigurationManager took ownership of all the providers except the ChainedConfigurationSource to itself.

What about the dangling change tokens?

Reply:
Since all the providers get disposed, I would expect the change tokens to be unregistered from any filewatchers or anything like that (and manual testing confirms that). We could do a [similar trick to HostBuilder](https://github.com/dotnet/runtime/blob/3fa2860847df903b448d0af228b3362ca72480db/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs#L285) and call `Services.GetService<IEnumerable<IConfiguration>>()` to resolve both the HostBuilder's ConfgiruationRoot and WebApplicationBuilder's ConfigurationManager out of DI and ensure both get disposed. This will double-dispose providers added by tests or IHostingStartups, but that should be safe.

Reply:
I did the `GetService<IEnumerable<IConfiguration>>()` thing: https://github.com/dotnet/aspnetcore/pull/37039/files#diff-f0e6f715f99a336de90bdbeef327cb71b7ee811d00d7c1722049557adaa009c5R227